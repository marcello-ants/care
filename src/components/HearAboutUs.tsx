/* eslint-disable no-param-reassign */
import {
  MenuItem,
  FormControl,
  FormHelperText,
  makeStyles,
  SelectProps,
  InputLabel,
} from '@material-ui/core';
import { Select } from '@care/react-component-lib';
import { useField } from 'formik';
import { HOW_DID_YOU_HEAR_ABOUT_US } from '../__generated__/globalTypes';

type HearAboutUsOptionsType = Array<{ value: HOW_DID_YOU_HEAR_ABOUT_US; label: string }>;

export const HearAboutUsOptions: HearAboutUsOptionsType = [
  { value: HOW_DID_YOU_HEAR_ABOUT_US.TV_AD, label: 'Cable TV Ad' },
  { value: HOW_DID_YOU_HEAR_ABOUT_US.SEARCH_ENGINE, label: 'Search Engine (Google, Bing)' },
  { value: HOW_DID_YOU_HEAR_ABOUT_US.PARENTING_GROUP, label: 'Parenting Group or Forum' },
  { value: HOW_DID_YOU_HEAR_ABOUT_US.FRIENDS_OR_FAMILY, label: 'Friends or Family' },
  { value: HOW_DID_YOU_HEAR_ABOUT_US.BANNER_AD, label: 'Banner Ad' },
  {
    value: HOW_DID_YOU_HEAR_ABOUT_US.PRESS_COVERAGE,
    label: 'Press Coverage (News, Magazine, Blog)',
  },
  { value: HOW_DID_YOU_HEAR_ABOUT_US.FACEBOOK, label: 'Facebook or Instagram' },
  { value: HOW_DID_YOU_HEAR_ABOUT_US.ONLINE_VIDEO, label: 'Streaming Video Ad (Hulu, Roku)' },
  { value: HOW_DID_YOU_HEAR_ABOUT_US.YOUTUBE, label: 'YouTube' },
  { value: HOW_DID_YOU_HEAR_ABOUT_US.BILLBOARD, label: 'Billboard' },
  {
    value: HOW_DID_YOU_HEAR_ABOUT_US.RADIO_AUDIO_AD,
    label: 'Radio/Audio Ad (iHeart, Pandora, Podcast)',
  },
  { value: HOW_DID_YOU_HEAR_ABOUT_US.INFLUENCER, label: 'Influencer' },
  {
    value: HOW_DID_YOU_HEAR_ABOUT_US.OTHER_SOCIAL_MEDIA,
    label: 'Other Social Media (Twitter, Pinterest, LinkedIn, TikTok)',
  },
  { value: HOW_DID_YOU_HEAR_ABOUT_US.OTHER, label: 'Other' },
];

//  Fisher-Yates Algorithm to shuffle arrays ramdomly
function shuffle(array: HearAboutUsOptionsType) {
  // removes last two items of the array
  // first puts after Facebook or Instagram option
  // last puts back at the end so that it doesn't get shuffled
  const otherOptions = array.slice(-2);
  const rest = array.slice(0, array.length - 2);

  // eslint-disable-next-line no-plusplus
  for (let i = rest.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * i);
    const temp = rest[i];
    rest[i] = rest[j];
    rest[j] = temp;
  }

  const fbIndex = rest.findIndex((elem) => elem.value === HOW_DID_YOU_HEAR_ABOUT_US.FACEBOOK);
  rest.splice(fbIndex + 1, 0, otherOptions[0]);

  return [...rest, otherOptions[1]];
}

const shuffledHearAboutUsOptions = shuffle(HearAboutUsOptions);

const useStyles = makeStyles((theme) => ({
  // a workaround needed until the DNA-1169 is done
  selectContainer: {
    width: '100%',
    paddingTop: theme.spacing(2.5),
    paddingLeft: theme.spacing(1),
    paddingRight: theme.spacing(1),
    marginBottom: theme.spacing(0),

    '& .MuiInput-input': {
      paddingTop: theme.spacing(2.25),
      paddingBottom: theme.spacing(0),
      paddingLeft: theme.spacing(2),
      color: theme.palette.care?.navy[900],
      fontSize: '18px',
      fontWeight: 700,
    },
    // styles for placeholder value
    '& .no-value .MuiInput-input': {
      color: theme.palette.care?.grey[700],
      fontWeight: 'normal',
    },
    '& .MuiFormHelperText-root': {
      marginTop: theme.spacing(0.5),
      paddingLeft: theme.spacing(0.25),
    },
  },
}));

interface HearAboutUsProps extends SelectProps {
  id?: string;
  name?: string;
  value?: string;
  helpText?: string;
  errorDisplayed?: string | boolean;
  withLabelText?: boolean;
}

const HearAboutUs = ({
  id,
  name,
  value,
  helpText,
  errorDisplayed,
  onChange,
  onBlur,
  withLabelText,
}: HearAboutUsProps) => {
  const classes = useStyles();

  return (
    <FormControl error={Boolean(errorDisplayed)} className={classes.selectContainer}>
      {withLabelText && <InputLabel id={id}>How did you hear about us?</InputLabel>}
      <Select
        id={id}
        name={name}
        value={value}
        onChange={onChange}
        onBlur={onBlur}
        displayEmpty
        className={!value ? 'no-value' : ''}
        inputProps={{
          tabIndex: -1,
        }}>
        <MenuItem value="" disabled>
          {withLabelText ? 'Select option' : 'How did you hear about us?'}
        </MenuItem>
        {shuffledHearAboutUsOptions.map((option) => (
          <MenuItem key={option.value} value={option.value}>
            {option.label}
          </MenuItem>
        ))}
      </Select>
      <FormHelperText>{helpText}</FormHelperText>
    </FormControl>
  );
};

HearAboutUs.defaultProps = {
  id: 'source',
  name: 'source',
  value: '',
  helpText: '',
  errorDisplayed: '',
  withLabelText: true,
};

interface HearAboutUsFieldProps {
  name: string;
  withLabelText?: boolean;
  required?: boolean;
}

function validateFieldRequired(value: string) {
  return HearAboutUsOptions.some((option) => option.value === value)
    ? undefined
    : 'Please select an option';
}

function validateFieldOptional(value: string) {
  if (!value) return undefined;

  return HearAboutUsOptions.some((option) => option.value === value)
    ? undefined
    : 'Please select an option';
}

export function HearAboutUsField({ name, required, withLabelText }: HearAboutUsFieldProps) {
  const validate = required ? validateFieldRequired : validateFieldOptional;
  const [field, meta] = useField({ name, validate });

  return (
    <HearAboutUs
      {...field}
      helpText={meta.error || ' '}
      errorDisplayed={meta.touched && meta.error}
      withLabelText={withLabelText}
    />
  );
}

HearAboutUsField.defaultProps = {
  required: false,
  withLabelText: true,
};

export default HearAboutUs;
