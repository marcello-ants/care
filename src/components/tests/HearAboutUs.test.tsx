import React from 'react';
import { render, screen, fireEvent, within, act } from '@testing-library/react';
import HearAboutUs, { HearAboutUsOptions } from '../HearAboutUs';

const mockValue = HearAboutUsOptions[0].value;
const mockHelpText = 'This is a helper text';
const mockOnChangeHandler = jest.fn();
const mockOnBlurHandler = jest.fn();

describe('HearAboutUs', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(
      <HearAboutUs
        name="source"
        id="source"
        value={mockValue}
        onChange={mockOnChangeHandler}
        onBlur={mockOnBlurHandler}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with the default text', () => {
    render(
      <HearAboutUs
        name="source"
        id="source"
        onChange={mockOnChangeHandler}
        onBlur={mockOnBlurHandler}
      />
    );
    expect(screen.getByText('How did you hear about us?')).toBeInTheDocument();
  });

  HearAboutUsOptions.forEach((currentOption) => {
    it(`should display ${currentOption.label} if passing ${currentOption.value}`, () => {
      render(
        <HearAboutUs
          name="source"
          id="source"
          value={currentOption.value}
          onChange={mockOnChangeHandler}
          onBlur={mockOnBlurHandler}
        />
      );

      expect(screen.getByText(currentOption.label)).toBeInTheDocument();
    });
  });

  it('should call the onChange and onBlur handlers', async () => {
    render(
      <HearAboutUs
        name="source"
        id="source"
        helpText={mockHelpText}
        onChange={mockOnChangeHandler}
        onBlur={mockOnBlurHandler}
      />
    );

    fireEvent.mouseDown(screen.getByRole('button'));
    const listbox = within(screen.getByRole('listbox'));
    fireEvent.click(listbox.getByText('Cable TV Ad'));

    await act(() => new Promise<undefined>((resolve) => setTimeout(resolve, 0)));

    expect(mockOnChangeHandler).toHaveBeenCalled();
    fireEvent.blur(screen.getByRole('button'));
    expect(mockOnBlurHandler).toHaveBeenCalled();
  });

  it('should show helper text', () => {
    render(
      <HearAboutUs
        name="source"
        id="source"
        value={mockValue}
        helpText={mockHelpText}
        onChange={mockOnChangeHandler}
        onBlur={mockOnBlurHandler}
      />
    );
    expect(screen.getByText(mockHelpText)).toBeInTheDocument();
  });
});
