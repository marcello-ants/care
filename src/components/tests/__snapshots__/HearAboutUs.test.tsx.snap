// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HearAboutUs should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root makeStyles-selectContainer-1"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink="true"
      id="source"
    >
      How did you hear about us?
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    >
      <div
        aria-haspopup="listbox"
        aria-labelledby="source"
        class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
        id="source"
        role="button"
        tabindex="-1"
      >
        Cable TV Ad
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput"
        name="source"
        tabindex="-1"
        value="TV_AD"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root makeStyles-root-4"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M18.2 8a.91.91 0 011.2 0 .91.91 0 010 1.2L12.6 16a.91.91 0 01-1.2 0L4.6 9.2a.91.91 0 010-1.2.91.91 0 011.2 0l6.2 6.2z"
          fill="#223f5f"
        />
        <path
          d="M12 16.37a1 1 0 01-.71-.26l-6.8-6.8a1.09 1.09 0 010-1.42 1.09 1.09 0 011.42 0L12 14l6.09-6.1a1.09 1.09 0 011.42 0 1.09 1.09 0 010 1.42l-6.8 6.8a1 1 0 01-.71.25zM5.2 7.93a.69.69 0 00-.49.18.76.76 0 000 1l6.8 6.8a.78.78 0 001 0l6.8-6.8a.76.76 0 000-1 .76.76 0 00-1 0L12 14.41l-.11-.1-6.2-6.2a.69.69 0 00-.49-.18z"
          fill="#223f5f"
        />
      </svg>
    </div>
    <p
      class="MuiFormHelperText-root MuiFormHelperText-filled"
    />
  </div>
</DocumentFragment>
`;
